File: golang/day16/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Direction = int

const (
	UP = iota
	RIGHT
	DOWN
	LEFT
)

func oppositeDirection(direction Direction) Direction {
	switch direction {
	case UP:
		return DOWN
	case DOWN:
		return UP
	case LEFT:
		return RIGHT
	case RIGHT:
		return LEFT
	}
	panic("Unknown Direction")
}

func loadData(filename string) ([][]rune, error) {
	data, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer func() {
		if err := data.Close(); err != nil {
			panic(err)
		}
	}()
	scanner := bufio.NewScanner(data)
	grid := make([][]rune, 0)
	for scanner.Scan() {
		grid = append(grid, []rune(scanner.Text()))
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return grid, nil
}

func padGrid(grid [][]rune) [][]rune {
	dim := len(grid) + 2
	newGrid := make([][]rune, dim)
	newGrid[0] = make([]rune, dim)
	for i := 0; i < dim; i++ {
		newGrid[0][i] = '#'
	}
	for i := 1; i < dim-1; i++ {
		newGrid[i] = append(newGrid[i], '#')
		newGrid[i] = append(newGrid[i], grid[i-1]...)
		newGrid[i] = append(newGrid[i], '#')
	}
	newGrid[dim-1] = make([]rune, dim)
	for i := 0; i < dim; i++ {
		newGrid[dim-1][i] = '#'
	}
	return newGrid
}

type Point struct {
	x int
	y int
}

func (p *Point) move(direction Direction) {
	switch direction {
	case UP:
		p.y -= 1
	case DOWN:
		p.y += 1
	case LEFT:
		p.x -= 1
	case RIGHT:
		p.x += 1
	}
}

type Status struct {
	Point
	direction Direction
}

func (s *Status) move() {
	s.Point.move(s.direction)
}

func (s *Status) moveBack() {
	s.Point.move(oppositeDirection(s.direction))
}

func processCorner(status *Status) {
	switch status.direction {
	case UP:
		status.direction = RIGHT
	case DOWN:
		status.direction = LEFT
	case LEFT:
		status.direction = DOWN
	case RIGHT:
		status.direction = UP
	}
}

func processBackwardCorner(status *Status) {
	switch status.direction {
	case UP:
		status.direction = LEFT
	case DOWN:
		status.direction = RIGHT
	case LEFT:
		status.direction = UP
	case RIGHT:
		status.direction = DOWN
	}
}

func processUpDown(status Status) []Status {
	if status.direction == LEFT || status.direction == RIGHT {
		statusOne := status
		statusOne.direction = UP
		statusTwo := status
		statusTwo.direction = DOWN
		return []Status{statusOne, statusTwo}
	}
	return []Status{status}
}

func processLeftRight(status Status) []Status {
	if status.direction == UP || status.direction == DOWN {
		statusOne := status
		statusOne.direction = LEFT
		statusTwo := status
		statusTwo.direction = RIGHT
		return []Status{statusOne, statusTwo}
	}
	return []Status{status}
}

func calculateEnergizedTiles(grid [][]rune, start Status) int {
	visited := make(map[Status]struct{})
	count := make(map[Point]struct{})
	queue := make([]Status, 0, 5)
	queue = append(queue, start)
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		if _, ok := visited[current]; ok {
			continue
		}
		visited[current] = struct{}{}
		count[current.Point] = struct{}{}
		current.move()
		tile := grid[current.y][current.x]
		switch tile {
		case '/':
			processCorner(&current)
			queue = append(queue, current)
		case '\\':
			processBackwardCorner(&current)
			queue = append(queue, current)
		case '|':
			queue = append(queue, processUpDown(current)...)
		case '-':
			queue = append(queue, processLeftRight(current)...)
		case '.':
			if grid[current.y][current.x] == '.' {
				count[current.Point] = struct{}{}
				current.move()
			}
			current.moveBack()
			queue = append(queue, current)
		}
	}
	return len(count) - 1
}

func solutionPart1() {
	grid, err := loadData("input.txt")
	grid = padGrid(grid)
	if err != nil {
		panic(err)
	}
	count := calculateEnergizedTiles(grid, Status{Point{x: 0, y: 1}, RIGHT})
	fmt.Println("Count:", count)
}

func solutionPart2() {
	grid, err := loadData("input.txt")
	grid = padGrid(grid)
	if err != nil {
		panic(err)
	}
	energizedTiles := make([]int, 0)
	for i := 1; i < len(grid)-1; i++ {
		energizedTiles = append(energizedTiles, calculateEnergizedTiles(grid, Status{Point{0, i}, RIGHT}))
		energizedTiles = append(energizedTiles, calculateEnergizedTiles(grid, Status{Point{len(grid) - 1, i}, LEFT}))
		energizedTiles = append(energizedTiles, calculateEnergizedTiles(grid, Status{Point{i, 0}, DOWN}))
		energizedTiles = append(energizedTiles, calculateEnergizedTiles(grid, Status{Point{i, len(grid) - 1}, UP}))
	}
	maxTiles := 0
	for _, el := range energizedTiles {
		maxTiles = max(maxTiles, el)
	}
	fmt.Println("Count:", maxTiles)
}

func main() {
	solutionPart1()
	solutionPart2()
}

File: Python/day16/main.py
import time
from collections import deque
from enum import Enum
from typing import Deque


class Direction(Enum):
    UP = 1
    DOWN = 2
    LEFT = 3
    RIGHT = 4


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def shift(self, direction: Direction):
        if direction == Direction.UP:
            self.y -= 1
        elif direction == Direction.DOWN:
            self.y += 1
        elif direction == Direction.LEFT:
            self.x -= 1
        elif direction == Direction.RIGHT:
            self.x += 1

    def __repr__(self):
        return f"Point({self.x}, {self.y})"

    def __str__(self):
        return f"({self.x}, {self.y})"

    def tuple(self) -> tuple[int, int]:
        return self.x, self.y



class Status:

    def __init__(self, point: Point, direction: Direction):
        self.point = point
        self.direction = direction

    def move(self):
        self.point.shift(self.direction)

    def __repr__(self):
        return f"Status({self.point.x}, {self.point.y}, {self.direction})"

    def __str__(self):
        return self.__repr__()

    def tuple(self) -> tuple[int, int, int]:
        return self.point.x, self.point.y, self.direction.value


def move_point(point: (int, int), direction: int) -> tuple[int, int, int]:
    if direction == Direction.UP.value:
        return point[0], point[1] - 1, direction
    elif direction == Direction.DOWN.value:
        return point[0], point[1] + 1, direction
    elif direction == Direction.LEFT.value:
        return point[0] - 1, point[1], direction
    elif direction == Direction.RIGHT.value:
        return point[0] + 1, point[1], direction


def load_data(filename: str) -> list[list[str]]:
    data = []
    with open(filename, 'r') as file:
        for line in file:
            directions = []
            for char in line.strip():
                directions.append(char)
            data.append(directions)
    return data


def pad_grid(grid: list[list[str]]) -> list[list[str]]:
    dim = len(grid) + 2
    new_grid = [['#' for _ in range(dim)]]
    for row in grid:
        row.insert(0, '#')
        row.append('#')
        new_grid.append(row)
    new_grid.append(['#' for _ in range(dim)])
    return new_grid


def process_splitter(status: tuple[int, int, int], queue: Deque[tuple[int, int, int]],
                     direction: tuple[Direction, Direction], new_direction: tuple[Direction, Direction]):
    if status[2] == direction[0].value or status[2] == direction[1].value:
        queue.append((status[0], status[1], new_direction[0].value))
        queue.append((status[0], status[1], new_direction[1].value))
    else:
        queue.append(status)


def process_corner(status: tuple[int, int, int], queue: Deque[tuple[int, int, int]],
                   new_direction: tuple[Direction, Direction, Direction, Direction]):
    if status[2] == Direction.UP.value:
        queue.append((status[0], status[1], new_direction[0].value))
    elif status[2] == Direction.DOWN.value:
        queue.append((status[0], status[1], new_direction[1].value))
    elif status[2] == Direction.LEFT.value:
        queue.append((status[0], status[1], new_direction[2].value))
    else:
        queue.append((status[0], status[1], new_direction[3].value))


def calculate_energized_tiles(grid: list[list[str]], start: Status) -> int:
    queue: Deque[tuple[int, int, int]] = deque([start.tuple()])
    visited: set[tuple[int, int, int]] = set()
    count: set[tuple[int, int]] = set()
    while len(queue) > 0:
        current = queue.popleft()
        if current in visited:
            continue
        visited.add(current)
        count.add((current[0], current[1]))
        current = move_point(current[:2], current[2])
        if grid[current[1]][current[0]] == '-':
            process_splitter(current, queue, (Direction.UP, Direction.DOWN), (Direction.LEFT, Direction.RIGHT))
        elif grid[current[1]][current[0]] == "|":
            process_splitter(current, queue, (Direction.LEFT, Direction.RIGHT), (Direction.UP, Direction.DOWN))
        elif grid[current[1]][current[0]] == "/":
            process_corner(current, queue, (Direction.RIGHT, Direction.LEFT, Direction.DOWN, Direction.UP))
        elif grid[current[1]][current[0]] == "\\":
            process_corner(current, queue, (Direction.LEFT, Direction.RIGHT, Direction.UP, Direction.DOWN))
        elif grid[current[1]][current[0]] == ".":
            if current[2] == Direction.DOWN.value:
                current_y = current[1]
                while grid[current_y + 1][current[0]] == ".":
                    count.add((current[0], current_y))
                    current_y += 1
                queue.append((current[0], current_y, current[2]))
            elif current[2] == Direction.UP.value:
                current_y = current[1]
                while grid[current_y - 1][current[0]] == ".":
                    count.add((current[0], current_y))
                    current_y -= 1
                queue.append((current[0], current_y, current[2]))
            elif current[2] == Direction.LEFT.value:
                current_x = current[0]
                while grid[current[1]][current_x - 1] == ".":
                    count.add((current_x, current[1]))
                    current_x -= 1
                queue.append((current_x, current[1], current[2]))
            else:  # Right
                current_x = current[0]
                while grid[current[1]][current_x + 1] == ".":
                    count.add((current_x, current[1]))
                    current_x += 1
                queue.append((current_x, current[1], current[2]))
    return len(count) - 1


def print_grid(energized_points: set[tuple[int, int]]):
    dim = 12
    grid = [['.' for _ in range(dim)] for _ in range(dim)]
    for point in energized_points:
        grid[point[1]][point[0]] = '#'
    for row in grid:
        print("".join(row))
    print()


def solution_part_1():
    grid = load_data('input.txt')
    grid = pad_grid(grid)
    start = Status(Point(0, 1), Direction.RIGHT)
    energized_tiles = calculate_energized_tiles(grid, start)
    print(energized_tiles)


def solution_part_2():
    grid = pad_grid(load_data("input.txt"))
    dim = len(grid)
    energized_tiles = []
    for i in range(dim - 2):
        energized_tiles.append(calculate_energized_tiles(
            grid, Status(Point(0, i + 1), Direction.RIGHT)
        ))
        energized_tiles.append(calculate_energized_tiles(
            grid, Status(Point(dim - 1, i + 1), Direction.LEFT)
        ))
        energized_tiles.append(calculate_energized_tiles(
            grid, Status(Point(i + 1, 0), Direction.DOWN)
        ))
        energized_tiles.append(calculate_energized_tiles(
            grid, Status(Point(i + 1, dim - 1), Direction.UP)
        ))
    print(max(energized_tiles))


def main():
    start = time.time()
    solution_part_1()
    solution_part_2()
    end = time.time()
    print(f"Time {(end - start)}s")


if __name__ == "__main__":
    main()
